File: leetcode/Algorithms/114.FlattenBinaryTreeToLinkedList/Solution.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Solution(object):
    def flatten(self, root):
        """
        :type root: TreeNode
        :rtype: void Do not return anything, modify root in-place instead.
        """
        if not root:
            return

        self.flatten(root.left)
        self.flatten(root.right)

        tail = root.left
        if tail:
            while tail.right:
                tail = tail.right

            tail.right = root.right
            root.right = root.left
            root.left = None

File: leetcode/Algorithms/114.FlattenBinaryTreeToLinkedList/Solution.java
import java.util.Stack;

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */

public class Solution {
    public void flatten(TreeNode root) {
        Stack<TreeNode> stack = new Stack<>();
        TreeNode cursor = root;
        while (cursor != null) {
            if (cursor.right != null) stack.push(cursor.right);
            if (cursor.left != null) {
                cursor.right = cursor.left;
            } else {
                if (!stack.isEmpty()) cursor.right = stack.pop();
            }
            cursor.left = null;
            cursor = cursor.right;
        }
    }
}
